### PR TITLE
Add Tenant Data Version and New Endpoint GET /schema/{className}/tenants/{tenantName}

### PR DESCRIPTION
**Pull request by nathanwilk7**
### What's being changed:

Adds a new tenant data version field, which starts at 0 and is incremented each time the tenant is frozen. For example, when a tenant is created with the status HOT, its data version starts at 0. When the tenant is frozen for the first time, then its data version will be 1, etc.

This data version field is passed around internally for most methods with read tenants, but the only place it is exposed externally is a new HTTP endpoint: `GET /schema/{className}/tenants/{tenantName}`. For now, we're considering this new field somewhat experimental until the querier is GA, so things may change, this is why I didn't add the field to all tenant endpoints.

We're not going to add this new endpoint and field to the client yet (the querier will just use direct HTTP for now), but may decide to add it to the clients later.

@kavirajk @reyreaud-l to review the tenant data version and raft integration
@moogacs to review the offload-related stuff and raft integration
@tsmith023 @dirkkul to review the API changes

### Review checklist

- [x] Documentation has been updated, if necessary. Link to changed documentation:
- [ ] Chaos pipeline run or not necessary. Link to pipeline:
- [x] All new code is covered by tests where it is reasonable.
- [ ] Performance tests have been run or not necessary.

### CODE REVIEW

File: cluster/schema/meta_class.go
Comment (kavirajk) [inline cluster/schema/meta_class.go:489]: what happens if multiple clients to offload the tenant at the same time? is it concurrent safe?

File: test/modules/offload-s3/offload_test.go
Comment (kavirajk) [inline test/modules/offload-s3/offload_test.go:194]: also better to lock the behavior of "multiple clients try to freeze the tenant concurrently" and still version is incremented once?

**Review by kavirajk — COMMENTED** [review]
LGTM. Nice work. few minor clarifications.

**Review by kavirajk — APPROVED** [review]

File: adapters/handlers/rest/handlers_schema.go
Comment (moogacs) [inline adapters/handlers/rest/handlers_schema.go:298]: **thought:**  do you think it will need authZ or this for internal usage ? 

i think it shall have `authZ` checks given that it's exposed to the open API

e.g.
```
h.authorizer.Authorize(principal, authorization.READ, authorization.Shards(params.ID)...)
```

**Review by moogacs — APPROVED** [review]
LGTM, left thought about AuthZ for the endpoint

File: adapters/handlers/rest/handlers_schema.go
Comment (nathanwilk7) [reply adapters/handlers/rest/handlers_schema.go:298]: Currently, that auth check happens inside the `s.manager.GetConsistentTenants` method call (same as in the existing `getTenants` method above). I assume that's to share the auth call among both the HTTP and gRPC handlers.

But separate from this PR I'm curious how we should ensure that the auth checks are not missed on any endpoints...

Comment (nathanwilk7) [reply adapters/handlers/rest/handlers_schema.go:298]: Currently, that auth check happens inside the s.manager.GetConsistentTenants method call (same as in the existing getTenants method above). I assume that's to share the auth call among both the HTTP and gRPC handlers.

But separate from this PR I'm curious how we should ensure that the auth checks are not missed on any endpoints...

File: test/modules/offload-s3/offload_test.go
Comment (nathanwilk7) [reply test/modules/offload-s3/offload_test.go:194]: good idea, thanks! i added checks to the `Test_ConcurrentFreezeUnfreeze` func below

**Comment by nathanwilk7** [comment]
if anyone gets the chance to approve this before i'm online, please feel free to merge it, thanks!

File: openapi-specs/schema.json
Comment (dirkkul) [inline openapi-specs/schema.json:5510]: This could be a TenantParameter type that could be reused for input and output?

Comment (kavirajk) [reply openapi-specs/schema.json:5510]: Sounds reasonable 👍. @nathanwilk7 something you can consider in the follow up PR?. I'm merging this one to unblock

**Review by kavirajk — APPROVED** [review]

**Review by reyreaud-l — APPROVED** [review]

File: openapi-specs/schema.json
Comment (kavirajk) [reply openapi-specs/schema.json:5510]: Ah couldn't merge because of this CI failure. 
Screenshot 2024-11-15 at 09 20 51

may be @dirkkul feedback could reduce this.

Comment (kavirajk) [reply openapi-specs/schema.json:5510]: I learned. it won't be a blocker :)

**Pull request merged**
